public/app/features/dashboard/dashgrid/PanelObserver.tsx

```diff
@@ -0,0 +1,192 @@
+
+  private updateScrollListenerCallback(remove = false) {
+    if (this.scroller) {
+      const empty = _.isEmpty(this.registry);
+      if (empty || remove) {
+        if (this.listener) {
+          console.log('Removing dashboard scroll listener');
+          this.scroller.removeEventListener('scroll', this.listener);
+          this.listener = null;
+        }
+      } else if (!this.listener) {
+        console.log('Adding dashboard scroll listener');
+        this.listener = this.updateVisibility.bind(this);
+        this.scroller.addEventListener('scroll', this.listener, {
+          capture: true,
+          passive: true,
+        });
+      }
+    }
+  }
+
+  // The Scroll callback
+  updateVisibility(force = false) {
+    if (_.isEmpty(this.registry)) {
+      this.updateScrollListenerCallback(true);
+    } else if (this.scroller) {
+      const top = Math.floor(this.scroller.scrollTop / 10) * 10; // check every 10 pixels
+      if (top !== this.lastChecked || force === true) {
+        this.lastChecked = top;
+        this.updateVisibilityProps(
+          this.scroller.scrollTop - PanelObserverScroll.MARGIN,
+          this.scroller.scrollTop + PanelObserverScroll.MARGIN + this.scroller.offsetHeight
+        );
+      }
+    }
+  }
+
+  // Check visibility for all elements
+  updateVisibilityProps(view_top: number, view_bottom: number) {
+    //console.log( 'CHECK In Viewport', view_top, view_bottom );
+    this.registry.forEach((element, panel) => {
+      const top = this.getTop(element);
+      const bottom = top + element.offsetHeight;
+      const vis = !(view_top > bottom || view_bottom < top);
+      //console.log( 'VIS', vis, panel.title, top, bottom );
+      if (panel.visible !== vis) {
+        panel.visible = vis;
+        panel.events.emit(PANEL_VISIBILITY_CHANGED_EVENT, vis);
+      }
+    });
+  }
+}
+
+export class PanelObserverIntersection implements PanelObserver {
+  observer: IntersectionObserver;
+
+  constructor() {
+    this.observer = new IntersectionObserver(this.callback.bind(this), {
+      root: null, // the viewport
+      rootMargin: '200px', // buffer by 100
+      threshold: 0, // any pixel
+    });
+  }
+
+  //---------------------------------------------------------
+  // API
+  //---------------------------------------------------------
+
+  dispose() {
+    this.observer.disconnect();
+  }
+
+  watch(e: HTMLElement, panel: PanelModel) {
+    if (e && panel) {
+      e['data-garfana-panel'] = panel;
+      this.observer.observe(e);
+    }
+  }
+
+  // Called externally on big change
+  check() {
+    // nothing?
+  }
+
+  //---------------------------------------------------------
+  // Internal
+  //---------------------------------------------------------
+
+  private callback(entries: IntersectionObserverEntry[]) {
+    entries.forEach(entry => {
+      const panel = entry.target['data-garfana-panel'];
+      if (panel.visible !== entry.isIntersecting) {
```

Comment:
If scrolling speed is fast, panel becomes !intersecting => intersecting => !intersecting in one callback.
So, it is better to check all entries before emit event, and emit event if final visible status is changed.

Reply:
Thanks for the review @mtanda!

I split this into another pull request that just does the visibility changes, but punts lazy loading...

Here is the change you recommended:
https://github.com/grafana/grafana/pull/12531/files#diff-5d8cd443562284b52967432ac7b6b0f2R183

Reply:
What was your general experience with `IntersectionObserver` vs `scroll` listener?